### ISSUE TITLE
Query analysis dies with "RuntimeError: generator raised StopIteration"

### ISSUE THREAD

**Issue by niccottrell** [rueckstiess/mtools#681]
### Expected behavior

Normal results, no crash

### Actual/current behavior

I'm getting `RuntimeError: generator raised StopIteration` when running `mloginfo --queries` or `mlogfilter mongo500.log --from "20190205"`. The full stack is:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/mtools/util/logfile.py", line 260, in __iter__
    raise e
  File "/usr/local/lib/python3.7/site-packages/mtools/util/logfile.py", line 248, in __iter__
    le = self.next()
  File "/usr/local/lib/python3.7/site-packages/mtools/util/logfile.py", line 217, in next
    raise StopIteration
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/mlogfilter", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/site-packages/mtools/mlogfilter/mlogfilter.py", line 355, in main
    tool.run()
  File "/usr/local/lib/python3.7/site-packages/mtools/mlogfilter/mlogfilter.py", line 332, in run
    for logevent in self.logfile_generator():
  File "/usr/local/lib/python3.7/site-packages/mtools/mlogfilter/mlogfilter.py", line 231, in logfile_generator
    for logevent in self.args['logfile'][0]:
RuntimeError: generator raised StopIteration
```

### Steps to reproduce the actual/current behavior

I've attached the mongo500.log but it appears normal.
- [mongo500.log](https://github.com/rueckstiess/mtools/files/2832221/mongo500.log)

### Environment

| Software         | Version
| ---------------- | -------
| mtools           |   1.5.3
| MongoDB server   |  Same result on old 3.2 logs as well as newer 3.4 logs 
| Operating system | macOS 10.14.3 (18D42)
| Python       | Via brew: 3.7.1 

`pip3 list` gives:
```
Package         Version
--------------- -------
mtools          1.5.3  
packaging       19.0   
pip             19.0.1 
pip-review      1.0    
psutil          5.5.0  
pymongo         3.7.2  
pyparsing       2.3.1  
python-dateutil 2.7.5  
setuptools      40.7.3 
six             1.12.0 
wheel           0.32.3 
```

**Comment by stennie** [comment]
@niccottrell I believe this is a consequence of [PEP 479](https://www.python.org/dev/peps/pep-0479/) which adds some backward-incompatible exception handling that is enabled by default in Python 3.7.

As at mtools 1.5.3 we have only been testing with Python 2.7 and 3.6, so it is likely that downgrading to Python 3.6.x will fix this error in the short term. Both of your example command lines worked fine for me with the provided log sample in Python 3.6 (although there are no queries to output via `mloginfo`).

Can you try with Python 3.6 and confirm if that resolves your issue?

Thanks,
Stennie

**Comment by supersujj** [comment]
python -V
Python 3.6.8

 [========================================] 98.9 % Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/mtools/util/logfile.py", line 260, in __iter__
    raise e
  File "/usr/local/lib/python3.7/site-packages/mtools/util/logfile.py", line 248, in __iter__
    le = self.next()
  File "/usr/local/lib/python3.7/site-packages/mtools/util/logfile.py", line 217, in next
    raise StopIteration
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/mloginfo", line 11, in <module>
    load_entry_point('mtools==1.5.3', 'console_scripts', 'mloginfo')()
  File "/usr/local/lib/python3.7/site-packages/mtools/mloginfo/mloginfo.py", line 95, in main
    tool.run()
  File "/usr/local/lib/python3.7/site-packages/mtools/mloginfo/mloginfo.py", line 90, in run
    section.run()
  File "/usr/local/lib/python3.7/site-packages/mtools/mloginfo/sections/query_section.py", line 65, in run
    for i, le in enumerate(logfile):
RuntimeError: generator raised StopIteration

**Comment by stennie** [comment]
@supersujj The `python` binary in your path may be 3.6.8, but the output in your traceback appears to be using Python 3.7 and encountering the same exception as reported on this issue. This would be expected given the [PEP 479 change](https://github.com/rueckstiess/mtools/issues/681#issuecomment-460958342) I mentioned in an earlier comment.

If you invoke `mloginfo --version`, this should report the version of the Python interpreter being used by the script. Reinstalling mtools in your Python 3.6.8 environment should resolve your error.

Regards,
Stennie

**Comment by niccottrell** [comment]
I managed to force downgrade to Python 3.6 in brew (had to uninstall 3.7 with no depedencies), and then reinstalled mtools to get the python mappings correct. This resolved the error, confirming a 3.7 problem.

**Comment by supersujj** [comment]
@stennie yes,thank is for help,i see the problem and fix it,now have another proble like #684

**Comment by kinghrothgar** [comment]
What makes this even more problematic is mtools appears to also not work inside of a virtualenv on OS X given the framework error:

```RuntimeError: Python is not installed as a framework. The Mac OS X backend will not be able to function correctly if Python is not installed as a framework.```

So it seems truly the only solution is to downgrade your system's python :(

**Comment by stennie** [comment]
@kinghrothgar You can use pyenv or virtualenv, but have to build Python with `--enable-framework` for `matplotlib` on macOS: https://github.com/rueckstiess/mtools/issues/587.

We use `tox` to run tests in `py27` and `py36` virtualenvs, and I'm using `pyenv` on macOS to switch between environments. Anaconda is another option.

Regards,
Stennie

**Comment by stennie** [comment]
This will be resolved in mtools 1.6.0 via #682, so closing as a duplicate.
